java/ql/src/experimental/Security/CWE/CWE-321/HardcodedJwtKey.qll

```diff
@@ -0,0 +1,156 @@
+/**
+ * Provides sources and sinks for detecting JWT token signing vulnerabilities.
+ */
+
+import java
+private import semmle.code.java.dataflow.FlowSources
+
+/** The Java class `com.auth0.jwt.JWT`. */
+class Jwt extends RefType {
+  Jwt() { this.hasQualifiedName("com.auth0.jwt", "JWT") }
+}
+
+/** The Java class `com.auth0.jwt.JWTCreator.Builder`. */
+class JwtBuilder extends RefType {
+  JwtBuilder() { this.hasQualifiedName("com.auth0.jwt", "JWTCreator$Builder") }
+}
+
+/** The Java class `com.auth0.jwt.algorithms.Algorithm`. */
+class Algorithm extends RefType {
+  Algorithm() { this.hasQualifiedName("com.auth0.jwt.algorithms", "Algorithm") }
+}
+
+/**
+ * The Java interface `com.auth0.jwt.interfaces.JWTVerifier` or it implementation class
+ * `com.auth0.jwt.JWTVerifier`.
+ */
+class JwtVerifier extends RefType {
+  JwtVerifier() {
+    this.hasQualifiedName(["com.auth0.jwt", "com.auth0.jwt.interfaces"], "JWTVerifier")
+  }
+}
+
+/** The secret generation method declared in `com.auth0.jwt.algorithms.Algorithm`. */
+class GetSecretMethod extends Method {
+  GetSecretMethod() {
+    this.getDeclaringType() instanceof Algorithm and
+    (
+      this.getName().substring(0, 4) = "HMAC" or
+      this.getName().substring(0, 5) = "ECDSA" or
+      this.getName().substring(0, 3) = "RSA"
+    )
+  }
+}
+
+/** The `require` method of `com.auth0.jwt.JWT`. */
+class RequireMethod extends Method {
+  RequireMethod() {
+    this.getDeclaringType() instanceof Jwt and
+    this.hasName("require")
+  }
+}
+
+/** The `sign` method of `com.auth0.jwt.JWTCreator.Builder`. */
+class SignTokenMethod extends Method {
+  SignTokenMethod() {
+    this.getDeclaringType() instanceof JwtBuilder and
+    this.hasName("sign")
+  }
+}
+
+/** The `verify` method of `com.auth0.jwt.interfaces.JWTVerifier`. */
+class VerifyTokenMethod extends Method {
+  VerifyTokenMethod() {
+    this.getDeclaringType() instanceof JwtVerifier and
+    this.hasName("verify")
+  }
+}
```

Comment:
@luchua-bc Why is this sink needed? Since, you already check if a token is signed at the time of token generation, checking for token verify calls seems unnecessary and would lead to multiple duplicate alerts for the same underlying cause.

Now, there may be instances where only the `verify` method is invoked and the code-base does not have a token generation module. This may happen when the code is expected to be run on the client side and the token is generated on the server side. But, even in this case, the alert should still be at the token generation point and not at the consumption point. 

Without this block, even though QL may not raise an alert in the code-base, it would raise an alert when it is run against the server code-base.

Reply:
Thank @porcupineyhairs for reviewing this PR.

Both `verify` and `sign` are sinks and they are independent. The modelling allows tokens to be generated on one side (`sign`) and verified on the other side (`verify`), which is exactly what you said. That is, no matter a code-base has one or both, it will be detected by the query.

I think the confusion may come from the following output:
```
| HardcodedJwtKey.java:25:23:25:31 | algorithm | HardcodedJwtKey.java:15:42:15:59 | "hardcoded_secret" : String | HardcodedJwtKey.java:25:23:25:31 | algorithm | $@ is used to sign a JWT token. | HardcodedJwtKey.java:15:42:15:59 | "hardcoded_secret" | Hardcoded String |
| HardcodedJwtKey.java:25:23:25:31 | algorithm | HardcodedJwtKey.java:19:49:19:54 | SECRET : String | HardcodedJwtKey.java:25:23:25:31 | algorithm | $@ is used to sign a JWT token. | HardcodedJwtKey.java:19:49:19:54 | SECRET | Hardcoded String |
```
This is not duplicate caused by handling both calls. Instead it is caused by that both
```
   private static final String SECRET = "hardcoded_secret";
```
and
```
Algorithm algorithm = Algorithm.HMAC256(SECRET);
```
are considered source of hardcoded strings. Probably we can enhance this by only taking hardcoded string in the assignment statement as the source, that is, only to have `private static final String SECRET = "hardcoded_secret";` as the source.

Hope this answers your question.

@luchua-bc